soc: xlnx: zynq7000: add FPU support

mausys opened the pull request:
According to its datasheet the Zynq 7000 has a VFPv3U FPU

mausys committed:
soc: xlnx: zynq7000: add FPU support

According to its datasheet the Zynq 7000 has a VFPv3 FPU

Signed-off-by: Simon Maurer <mail@maurer.systems>

mausys commented:
datasheet:
https://docs.amd.com/v/u/en-US/ds190-Zynq-7000-Overview

nordicjm commented on cmake/gcc-m-fpu.cmake:
```diff
@@ -45,7 +45,19 @@ if("${ARCH}" STREQUAL "arm")
     set(FPU_FOR_cortex-m85+nodsp    auto)
 
     set(GCC_M_FPU ${FPU_FOR_${GCC_M_CPU}})
+  elseif(CONFIG_CPU_AARCH32_CORTEX_A)
```
please update `subsys/mgmt/mcumgr/grp/os_mgmt/include/os_mgmt_processor.h` also

mausys commented:
i checked the FPU on the HW with this function:
```
static void fpu_info(void)
{
      int val_fpsid = 0;
      int val_mvfr0 = 0;
      int val_mvfr1 = 0;

      __asm__ ("vmrs  %[output], FPSID"  : [output] "=r" (val_fpsid));
      __asm__ ("vmrs  %[output], MVFR0"  : [output] "=r" (val_mvfr0));
      __asm__ ("vmrs  %[output], MVFR1"  : [output] "=r" (val_mvfr1));

      LOG_INF("fpsid=0x%x mvfr0=0x%x mvfr1=0x%x\n", val_fpsid, val_mvfr0, val_mvfr1);
}
```
output: 
fpsid=0x41033094 mvfr0=0x10110222 mvfr1=0x1111111

So according to the reference manual (https://developer.arm.com/documentation/ddi0487/latest/) fp16 hw conversion is supported, but not the floating point exception trapping. But does it matter? VFP_FEATURE_TRAP isn't used anywhere.

mausys commented:
I removed the half-precision support. According to MVFR1 register only conversion and not arithmetic is supported and I'm not sure if for the gcc -mfpu=vfpv3-fp16  flag both must be supported.

mausys committed:
soc: xlnx: zynq7000: remove FPU FMAC support

Floating-Point System ID register (FPSID)        = 0x41033094
AArch32 Media and VFP Feature Register 0 (MVFR0) = 0x10110222
AArch32 Media and VFP Feature Register 0 (MVFR1) =  0x1111111

MVFR1 SIMDFMAC, bits [31:28] = 0; FMAC is not supported

Signed-off-by: Simon Maurer <mail@maurer.systems>

mausys commented:
btw wouldn't it be simpler if the SOC just selects the FPU architecture that can directly be used as gcc flag, instead of selecting the FPU features in the SOC Kconfig and guessing the FPU architecture in gcc-m-fpu.cmake?

e.g. select FPU in the SOC Kconfig:

`select FPU_ARCH_VFPV3`

and in Kconfig.vfp

```
config FPU_ARCH_VFPV3
	bool
	select CPU_HAS_VFP
	select VFP_FEATURE_SINGLE_PRECISION
	select VFP_FEATURE_DOUBLE_PRECISION
	select VFP_FEATURE_REGS_S64_D32
```
Also the VFP_FEATURE_FMAC/VFP_FEATURE_FMAC/VFP_FEATURE_TRAP/VFP_FEATURE_HALF_PRECISION aren't used in zephyr, only the number of FPU registers are important for a thread context switch.
